### ISSUE TITLE
Full crashes and doubled notifications

### ISSUE THREAD

**Issue by gyvess** [GiedriusS/TwitchNotifier#7]
1) I used to run Twitchnotifier in the background and it'd mysteriously crash, so this time I ran it in terminal to monitor errors. It took days but it finally crashed and this is what it spit out. There are dozens of names within the redacted brackets, but I know they're all valid.
```
Traceback (most recent call last):
  File "/usr/bin/twitchnotifier", line 104, in 
      API.diff(API.get_status())
        File "/usr/lib/python3.8/site-packages/libtn.py", line 316, in get_status
            chan_ids = self.get_userids(chans)
              File "/usr/lib/python3.8/site-packages/libtn.py", line 184, in get_userids
                  raise NameError(f'{nicks} has invalid nicknames')
                  NameError: [ redacted ] has invalid nicknames
```

2) I've been using the `-i 30` argument and in recent months the notifications have been showing streamers coming online, then going offline, and finally back online in quick succession. This may be the fault of a Twitch API change because I've also somewhat observed this in another application. Other than the previous crash causing error, here are the other received error messages that may be relevant to this problem.
```
Exception in access_kraken::requests.get()
__doc__ = A Connection error occurred.
Failed to parse json in access_kraken
```

**Comment by GiedriusS** [comment]
Hi! Thank you for using TN and trying to dig into this. This is interesting because we have:

https://github.com/GiedriusS/TwitchNotifier/blob/master/libtn.py#L218-L220
https://github.com/GiedriusS/TwitchNotifier/blob/master/libtn.py#L234-L239

and we check for `None` here: 

https://github.com/GiedriusS/TwitchNotifier/blob/master/libtn.py#L183

Do these errors occur _before_ it crashes with that exception? Does it show anything interesting with `--verbose`?

**Comment by gyvess** [comment]
Apologies for not replying sooner. Yes, the second set of errors do occur multiple times prior to the error given during the crash. I now have twitchnotifier running with the `--verbose` argument and the errors are being thrown into a log. I'm just waiting for it to crash again so I can give you better information. 

In the meantime, are you open to feature requests?

**Comment by gyvess** [comment]
The log file is over 5GB in size but it finally happened again. Here are the lines that comprise the tail end of the file, which diverge from what is typically spit out prior to the crash.

```
cmd: /users, payload: {'login': ''}
req.text: {"error":"Bad Request","status":400,"message":"The login parameter is invalid: \"\""}
req.status_code: 400
req.headers: {'Connection': 'keep-alive', 'Content-Length': '85', 'Content-Type': 'application/json; charset=utf-8', 'Access-Control-Allow-Origin': '*', 'Cache-Control': 'no-cache, no-store, must-revalidate, private', 'Expires': '0', 'Pragma': 'no-cache', 'Timing-Allow-Origin': 'https://www.twitch.tv', 'Date': 'Mon, 28 Sep 2020 22:45:48 GMT', 'X-Served-By': 'cache-sea4461-SEA, cache-mia17632-MIA', 'X-Cache': 'MISS, MISS', 'X-Cache-Hits': '0, 0', 'X-Timer': 'S1601333149.523589,VS0,VS0,VE86', 'Vary': 'Accept-Encoding', 'Strict-Transport-Security': 'max-age=300'}
--------------------
Kraken request returned bad code 400, bailing
Traceback (most recent call last):
  File "/usr/bin/twitchnotifier", line 104, in <module>
    API.diff(API.get_status())
  File "/usr/lib/python3.8/site-packages/libtn.py", line 316, in get_status
    chan_ids = self.get_userids(chans)
  File "/usr/lib/python3.8/site-packages/libtn.py", line 184, in get_userids
    raise NameError(f'{nicks} has invalid nicknames')
NameError: [] has invalid nicknames
```

**Comment by GiedriusS** [comment]
Sorry for the delay in response. Thanks for the report! It seems like that function assumes a non-empty array and then crashes when this https://github.com/GiedriusS/TwitchNotifier/blob/master/libtn.py#L315 gives an empty array. Need to check a bit more deeply. I am open to feature requests, feel free to propose anything.

**Comment by gyvess** [comment]
Beginning about a week ago Twitchnotifier no longer works with the `-c` argument. With the `--verbose` argument added I can see that it's essentially in a constant loop. My environment has remained the same as I haven't updated my system whatsoever so I'm wondering if it's just me. Again, this is when using the `-c` argument. I tested the `-u` argument and that still functions as expected.

As far as features, there are only two that I'd like. One is an additional argument for the `-c` option to only send notifications when streams come online, omitting offline notifications. Second, it'd be nice if you could send the streamer's profile picture to be used as an icon for the notifications. Watching what gets spit out of `--verbose` I noticed the word logo followed by a link to a crawled image, and that's the profile picture.

**Comment by GiedriusS** [comment]
> Beginning about a week ago Twitchnotifier no longer works with the `-c` argument. With the `--verbose` argument added I can see that it's essentially in a constant loop. My environment has remained the same as I haven't updated my system whatsoever so I'm wondering if it's just me. Again, this is when using the `-c` argument. I tested the `-u` argument and that still functions as expected.

Yep, I was able reproduce this as well. I will work on this very soon.

> 
> As far as features, there are only two that I'd like. One is an additional argument for the `-c` option to only send notifications when streams come online, omitting offline notifications. Second, it'd be nice if you could send the streamer's profile picture to be used as an icon for the notifications. Watching what gets spit out of `--verbose` I noticed the word logo followed by a link to a crawled image, and that's the profile picture.

Those are definitely easily doable! Could you please create another issue in this repo with this comment so that it wouldn't get lost?

**Comment by GiedriusS** [comment]
AFAICT it seems like they broke pagination on the v5 Twitch API :( Will try to ping them.

**Comment by GiedriusS** [comment]
@gyvess managed to get a quick response. Here is the issue for tracking the infinite loop problem: https://github.com/twitchdev/issues/issues/237. I have implemented a fix to the "infinite loop" problem here: https://github.com/GiedriusS/TwitchNotifier/pull/8.

**Comment by gyvess** [comment]
Last night when I was trying to think of factors I hadn't considered, I realized that two weeks had passed and I had yet to crash. Not only that, but it looks like the quick online, offline, and then online notifications have also stopped. Then as I was typing this reply `twitchnotifier` crashed. The error looks nearly exactly the same except a line number difference due to your recent changes and the brackets after "NameError:" were empty instead of listing tons of stream names.

```
Kraken request returned bad code 400, bailing
Traceback (most recent call last):
  File "/usr/bin/twitchnotifier", line 104, in 
      API.diff(API.get_status())
        File "/usr/lib/python3.8/site-packages/libtn.py", line 321, in get_status
            chan_ids = self.get_userids(chans)
              File "/usr/lib/python3.8/site-packages/libtn.py", line 184, in get_userids
                  raise NameError(f'{nicks} has invalid nicknames')
                  NameError: [] has invalid nicknames
```
My system has not changed with the exception of updating `twitchnotifier` from git. All other software remains the same.

As far is factors I hadn't considered, there are a few things. 
* I modified `twitchnotifier.cfg` by appending lines 19 and 31 with `(${%r})` in order to print in 12 hour time. It works as expected, but I'm not familiar with the python time module and simply retrieved this value from the `date` GNU coreutil. Listing this just in case this is somehow a bad practice.
* There is one user I follow named maitoz who's online notification differs from all others. It's as if the bold title of the notification is skipped and only the body is used. The only difference between him and most other streamers is his language setting has a hyphen in it, en-gb. All other streamers I follow are en, ko, or ja as far as I know. You might think the exclamation marks could be the culprit, but others streamers have those in their title as well and without issue.

    * As an example, here are pictures of what is normal notification behavior and how maitoz coming online is displayed. Note the top three notifications I originally believed to be doubling behavior, but on further inspection it seems to be streamer error. I'm using `dunst` for notifications by the way.
![image](https://user-images.githubusercontent.com/6826350/97332720-28bfad00-1851-11eb-97f8-918ea3926a2f.png)
![image](https://user-images.githubusercontent.com/6826350/97333783-4e00eb00-1852-11eb-97a4-73f4c5ebbd39.png)



* Continuing on the language thought line, I realize that a number of streamers I follow have english URLs but Korean/Japanese display names. By a similar token, there are also streamers who previously had one name only to have it changed it. This type of thing should be handled internally by twitch itself, but it's just something to consider given part of the error complains about "invalid nicknames." However, in the most recent crash I checked to see if my list of streamers online or offline had changed but it hadn't. So I dunno.

Anyway, that's all for now. One problem seemingly solved while the other remains persistent. I guess for the moment I can create a script that instantly relaunches `twitchnotifier` upon crashing.

**Comment by gyvess** [comment]
I keep forgetting to close this. 

A couple months back now I was banned from Twitch. I appealed and after some time they unbanned me stating that they mistakenly took my account for a bot. On my account I rarely talked in chat and definitely haven't said anything in half a decade, meaning I solely use it for notifications. Anyway, after being unbanned I haven't had a single crash or doubled notification. I'm going to assume that both errors were caused by Twitch limiting my account in someway prior to the ban.
